### ISSUE TITLE
exception : javax.net.ssl.SSLHandshakeException: Remote host closed connection during handshake

### ISSUE THREAD

**Issue by mwkirk** [mwkirk/javapns#32]
_Original author: cery...@gmail.com (October 14, 2010 08:32:40)_

<b>What steps will reproduce the problem?</b>
1.public class IphoneNotificationManager {

```
public static void sentMessage(NotificationDetail notificationDetail) {
    PayLoad aPayload = payLoadValueOf(notificationDetail);

    PushNotificationManager pushManager = PushNotificationManager
            .getInstance();
    // Get iPhone client
    try {
        pushManager.addDevice(notificationDetail.getDeviceId(),
                notificationDetail.getUuid());

        Device client = pushManager.getDevice(notificationDetail
                .getDeviceId());
        // Send message
        pushManager.sendNotification(client, aPayload);

        System.out.println(&quot;Message sent! &quot; + aPayload);
        System.out.println(&quot;# of attempts: &quot;
                + pushManager.getRetryAttempts());
        pushManager.removeDevice(notificationDetail.getDeviceId());
    } catch (Exception e) {
        System.out.println(&quot;exception : &quot; + e);
        e.printStackTrace();
    }
}

public static void initializeConnection() {
    PushNotificationManager pushManager = PushNotificationManager
            .getInstance();
    // Initialize connection
    try {
        pushManager.initializeConnection(ApplicationProperties.APPLE_HOST,
                ApplicationProperties.APPLE_PORT,
                ApplicationProperties.CERTIFICATE_FILE_PATH,
                ApplicationProperties.APPLE_PWD,
                SSLConnectionHelper.KEYSTORE_TYPE_PKCS12);
        System.out.println(&quot;connection establishd to APNS&quot;);
    } catch (Exception e) {
        System.out.println(&quot;not able to initialize connection to APNS&quot;);
    }
}

public static void stopConnection() {
    PushNotificationManager pushManager = PushNotificationManager
            .getInstance();
    try {
        pushManager.stopConnection();
        System.out.println(&quot;connection stopped succesfully&quot;);
    } catch (IOException e) {
        System.out.println(&quot;not able to stop connection to APNS&quot;);
    }
}

private static PayLoad payLoadValueOf(NotificationDetail notificationDetail) {
    PayLoad aPayload = null;
    try {
        aPayload = new PayLoad();
        aPayload.addAlert(notificationDetail.getMatchbetween() + &quot; at &quot;
                + notificationDetail.getMatchdate() + &quot; on &quot;
                + notificationDetail.getMatchtime());
        aPayload.addSound(&quot;default&quot;);

    } catch (JSONException e) {
        System.out.println(&quot;Exceptio : &quot; + e);
    }
    return aPayload;
}

public static void main(String args[]) {
    NotificationDetail notificationDetail = new NotificationDetail();
    notificationDetail
            .setUuid(&quot;ebbf8c927bfd6754348fb7ccf17cbe73f2e4741fc10f08fc2316ae877867eb80&quot;);
    notificationDetail.setMatchbetween(&quot;abc vs axy&quot;);
    notificationDetail.setDeviceId(&quot;1&quot;);
    notificationDetail.setMatchdate(&quot;today &quot;);
    notificationDetail.setMatchtime(&quot;12:30&quot;);
    initializeConnection();
    IphoneNotificationManager.sentMessage(notificationDetail);
    stopConnection();
}
```

}
1. Working fine on Window 7 but not working in linux server.

<b>What is the expected output? What do you see instead?</b>
Getting an exception 
exception : javax.net.ssl.SSLHandshakeException: Remote host closed connection during handshake
javax.net.ssl.SSLHandshakeException: Remote host closed connection during handshake
    at com.sun.net.ssl.internal.ssl.SSLSocketImpl.readRecord(SSLSocketImpl.java:801)
    at com.sun.net.ssl.internal.ssl.SSLSocketImpl.performInitialHandshake(SSLSocketImpl.java:1089)
    at com.sun.net.ssl.internal.ssl.SSLSocketImpl.writeRecord(SSLSocketImpl.java:618)
    at com.sun.net.ssl.internal.ssl.AppOutputStream.write(AppOutputStream.java:59)
    at java.io.OutputStream.write(OutputStream.java:58)
    at javapns.back.PushNotificationManager.sendNotification(Unknown Source)
    at com.src.IphoneNotificationManager.sentMessage(IphoneNotificationManager.java:29)
    at com.src.PushJob.execute(PushJob.java:20)
    at org.quartz.core.JobRunShell.run(JobRunShell.java:216)
    at org.quartz.simpl.SimpleThreadPool$WorkerThread.run(SimpleThreadPool.java:549)
Caused by: java.io.EOFException: SSL peer shut down incorrectly
    at com.sun.net.ssl.internal.ssl.InputRecord.read(InputRecord.java:333)
    at com.sun.net.ssl.internal.ssl.SSLSocketImpl.readRecord(SSLSocketImpl.java:782)
    ... 9 more

<b>What version of the product are you using? On what operating system?</b>
Using 
bcprov-jdk15-145.jar
javapns-jdk1.5-v1.6.2.jar

<b>Please provide any additional information below.</b>

_Original issue: http://code.google.com/p/javapns/issues/detail?id=32_

**Comment by mwkirk** [comment]
_From idbill.p...@gmail.com on October 16, 2010 21:34:30_
When you:
Enable SSL logging in java by adding the following to your java VM arguments:

-Djavax.net.debug=all

What SSL errors do you get?

**Comment by mwkirk** [comment]
_From cery...@gmail.com on October 19, 2010 13:36:46_
It works fine when we restart the tomcat server.

It appears again after sometime. Any help would be appreciated

**Comment by mwkirk** [comment]
_From idbill.p...@gmail.com on October 19, 2010 16:11:30_
Sounds like you are trying to maintain a socket connection or running into the 200 message limit that others have commented on. 

Would you verify?

**Comment by mwkirk** [comment]
_From cery...@gmail.com on October 20, 2010 11:27:55_
I tried sending around 300 push in a loop and they all are successful.

This above piece of code is called by a scheduled job through quartz. 
It works fine for 2-3 hrs. But after that starts giving this exception, everything
resolves after we restart the server.

**Comment by mwkirk** [comment]
_From idbill.p...@gmail.com on October 20, 2010 17:21:51_
Sounds like you are trying to maintain a socket connection.

The error is: Remote host closed connection during handshake

which would indicate that connection lapsed or you're not initializing the connection properly.

**Comment by mwkirk** [comment]
_From cery...@gmail.com on October 21, 2010 05:12:28_
I am calling above methods in this order
1. initializeConnection();
2. sentMessage(notificationDetail);
3. stopConnection();

But exception is coming in 2. call when we use sendNotification method of PushNotificationManager instance. 

I even tried with not calling 3. stopConnection as mentioned in some threads which i found. But that is also of no help. I am receiving same exception. 

I am not able to find any clue at all why this is happening??

**Comment by mwkirk** [comment]
_From idbill.p...@gmail.com on October 21, 2010 19:10:37_
It appears the getInstance is not syncronized. I'm updating the code to handle this.

Basically, when you call calling PushNotificationManager.getInstance you are getting a different instance of the singleton.

At the moment you would be better off having a class variable that stores that instance, so that you are guarnteed to get the same instance back.

Or you can try the new 163 version I just uploaded and provide feedback.

**Comment by mwkirk** [comment]
_From cery...@gmail.com on October 22, 2010 05:14:21_
Thanks for that i'll try that and update you on this.......

**Comment by mwkirk** [comment]
_From cery...@gmail.com on October 22, 2010 13:25:40_
I tried getting instance only once. But same error occurred again after some tries.

**Comment by mwkirk** [comment]
_From nikola.p...@gmail.com on November 28, 2010 23:56:32_
I was receiving the same javax.net.ssl.SSLHandshakeException exception and I solve this issue by removing all duplicates of org.json.JSONObject classes from my project (refactoring some packages helped ie org.json -> org.json.myframework). Javapns and otherframework included same org.json package with same class names within them - so that was the problem in my case - JAR hell :(

If you use Eclipse try Ctrl+Shift+T <JSONObject> to check for duplications.

**Comment by mwkirk** [comment]
_From cery...@gmail.com on November 29, 2010 05:18:53_
Thanks for replying.

I have checked that also, but there is no duplication of jars for org.json.JSONObject classes. There exist only one jar for these classes i.e. javapns-jdk1.5-v1.6.2.jar.

Regards

**Comment by mwkirk** [comment]
_From sathiyar...@gmail.com on April 17, 2011 15:44:09_
hi, I' getting the same error, please help me out

javax.net.ssl.SSLHandshakeException: Remote host closed connection during handshake
        at com.sun.net.ssl.internal.ssl.SSLSocketImpl.readRecord(SSLSocketImpl.java:808)
        at com.sun.net.ssl.internal.ssl.SSLSocketImpl.performInitialHandshake(SSLSocketImpl.java:1120)
        at com.sun.net.ssl.internal.ssl.SSLSocketImpl.writeRecord(SSLSocketImpl.java:623)
        at com.sun.net.ssl.internal.ssl.AppOutputStream.write(AppOutputStream.java:59)
        at java.io.OutputStream.write(OutputStream.java:58)
        at com.apple.ist.retail.pos.tr.protocol.tr.transport.Call.executeCall(Call.java:197)
        at com.apple.ist.retail.pos.tr.protocol.tr.transport.Transport.executeCall(Transport.java:210)
        at com.apple.ist.retail.pos.tr.protocol.tr.TenderRetailServiceImpl.authorizeDebitPurchase(TenderRetailServiceImpl.java:486)
        at com.extendyourstore.domain.manager.tenderauth.tenderretail.TRAuthTechnician.requestAuthorization(TRAuthTechnician.java:943)
        at com.extendyourstore.domain.manager.tenderauth.TenderAuthValet.execute(TenderAuthValet.java:107)
        at com.extendyourstore.domain.manager.tenderauth.tenderretail.TRAuthTechnician.transport(TRAuthTechnician.java:1066)
        at com.extendyourstore.domain.manager.tenderauth.NxsTenderAuthManager.transportValet(NxsTenderAuthManager.java:283)
        at com.extendyourstore.domain.manager.tenderauth.NxsTenderAuthManager.requestAuthorization(NxsTenderAuthManager.java:172)
        at com.extendyourstore.pos.services.tender.authorization.can.AuthorizeChargeSite.arrive(AuthorizeChargeSite.java:587)
        at com.extendyourstore.foundation.tour.engine.Site.arrive(Site.java:177)
        at com.extendyourstore.foundation.tour.engine.Road.traverseForward(Road.java:833)
        at com.extendyourstore.foundation.tour.engine.Road.traverse(Road.java:582)
        at com.extendyourstore.foundation.tour.engine.SiteVertex.driveSite(SiteVertex.java:285)
        at com.extendyourstore.foundation.tour.engine.AbstractSite.driveSite(AbstractSite.java:71)
        at com.extendyourstore.foundation.tour.engine.Site.driveSite(Site.java:153)
        at com.extendyourstore.foundation.tour.service.BusDriver.processLetter(BusDriver.java:315)
        at com.extendyourstore.foundation.tour.service.BusDriver.processLetters(BusDriver.java:213)
        at com.extendyourstore.foundation.tour.service.AbstractCrewMember.run(AbstractCrewMember.java:231)
        at com.extendyourstore.foundation.manager.thread.StaticCacheThread.run(StaticCacheThread.java:148)
Caused by: java.io.EOFException: SSL peer shut down incorrectly

**Comment by mwkirk** [comment]
_From idbill.p...@gmail.com on April 17, 2011 16:30:56_
My guess is that you are not connecting to the correct host and/or port.

What troubleshooting have you done

**Comment by mwkirk** [comment]
_From sype...@gmail.com on September 09, 2011 19:05:56_
ceryash, were you able to fix the issue you reported?

**Comment by mwkirk** [comment]
_From sype...@gmail.com on September 14, 2011 16:08:33_
Closing issue as no feedback was received by original poster for many months.  
Please feel free to re-open if the issue is still valid.
